cw_decred/lib/wallet.dart

```diff
@@ -0,0 +1,604 @@
+    );
+  }
+
+  @override
+  void setExceptionHandler(void Function(FlutterErrorDetails) onError) =>
+      onError;
+
+  Future<void> renameWalletFiles(String newWalletName) async {
+    final currentDirPath =
+        await pathForWalletDir(name: walletInfo.name, type: type);
+
+    final newDirPath = await pathForWalletDir(name: newWalletName, type: type);
+
+    if (File(newDirPath).existsSync()) {
+      throw "wallet already exists at $newDirPath";
+    }
+    ;
+
+    await Directory(currentDirPath).rename(newDirPath);
+  }
+
+  @override
+  Future<String> signMessage(String message, {String? address = null}) {
+    if (watchingOnly) {
+      throw "a watching only wallet cannot sign";
+    }
+    var addr = address;
+    if (addr == null) {
+      addr = libdcrwallet.currentReceiveAddress(walletInfo.name);
+    }
+    if (addr == null) {
+      throw "unable to get an address from unsynced wallet";
+    }
+    return libdcrwallet.signMessageAsync(
+        walletInfo.name, message, addr, _password);
+  }
+
+  List<Unspent> unspents() {
+    final res = libdcrwallet.listUnspents(walletInfo.name);
+    final decoded = json.decode(res);
+    var unspents = <Unspent>[];
+    for (final d in decoded) {
+      final spendable = d["spendable"] ?? false;
+      if (!spendable) {
+        continue;
+      }
+      final amountDouble = d["amount"] ?? 0.0;
+      final amount = (amountDouble * 1e8).toInt().abs();
+      final utxo = Unspent(
+          d["address"] ?? "", d["txid"] ?? "", amount, d["vout"] ?? 0, null);
+      utxo.isChange = d["ischange"] ?? false;
+      unspents.add(utxo);
+    }
+    this.updateUnspents(unspents);
+    return unspents;
+  }
+
+  void updateUnspents(List<Unspent> unspentCoins) {
+    if (this.unspentCoinsInfo.isEmpty) {
+      unspentCoins.forEach((coin) => this.addCoinInfo(coin));
+      return;
+    }
+
+    if (unspentCoins.isEmpty) {
+      this.unspentCoinsInfo.clear();
+      return;
+    }
+
+    final walletID = idPrefix + walletInfo.name;
+    if (unspentCoins.isNotEmpty) {
+      unspentCoins.forEach((coin) {
+        final coinInfoList = this.unspentCoinsInfo.values.where((element) =>
+            element.walletId == walletID &&
+            element.hash == coin.hash &&
+            element.vout == coin.vout);
+
+        if (coinInfoList.isEmpty) {
+          this.addCoinInfo(coin);
+        }
+      });
+    }
+
+    final List<dynamic> keys = <dynamic>[];
+    this.unspentCoinsInfo.values.forEach((element) {
+      final existUnspentCoins =
+          unspentCoins.where((coin) => element.hash.contains(coin.hash));
+
+      if (existUnspentCoins.isEmpty) {
+        keys.add(element.key);
+      }
+    });
+
+    if (keys.isNotEmpty) {
+      unspentCoinsInfo.deleteAll(keys);
+    }
+  }
+
+  void addCoinInfo(Unspent coin) {
+    final newInfo = UnspentCoinsInfo(
+      walletId: idPrefix + walletInfo.name,
+      hash: coin.hash,
+      isFrozen: false,
+      isSending: false,
```

Comment:
I thought isSending meant that the user always wants to use this input in a transaction. Are isSending a pool we should choose from? Then if isSending is false is the same as isFrozen?

Reply:
Doing as you say for now, but I think decred is currently handling this wrong. Right now it is absolutely using all `isSending` and then uses other utxo by selecting the least number that can make the amount we want to send plus fee. It doesn't use `isFrozen` at all.

Reply:
Ok, I think I got it now. `isSending` means we can use but we dont have to. If that is not checked, it is the same as `isFrozen`

Reply:
And now Decred handles it accordingly?
i.e
for Monero, we pass a list of preferred inputs (which are the ones with `isSending = true`) so that the Monero code uses from this list of inputs while creating the transaction

Reply:
If isSending == false means the same as outputs are frozen, then yes we are doing it correctly. It is only choosing from the `isSending`, i.e., it ignores everything else.

Well, it could be better, since we are telling the wallet which inputs not to use, if I had understood `isSending` better then it would just accept a list of inputs to use. Can we change that later if it seems to be a problem? I can imagine an edge case where a new transaction comes in while creating a transaction, and a blocked is mined at the same time, then we might include that utxo in the transaction although, it would be isSending == true by default, but the user would not have seen it in the list.